src/chain/chain.rs

```diff
@@ -996,29 +996,44 @@ impl Chain {
         let is_connected = |target_name: &XorName| connected_peers.contains(&target_name);
 
         let candidates = |target_name: &XorName| {
-            let mut filtered_sections =
+            let filtered_sections =
                 self.closest_sections(target_name)
                     .into_iter()
-                    .map(|(_, members)| {
+                    .map(|(prefix, members)| {
                         (
+                            prefix,
                             members.len(),
                             members.into_iter().filter(is_connected).collect::<Vec<_>>(),
                         )
                     });
-            let best_section = filtered_sections
-                .find(|(len, connected)| connected.len() * 3 >= *len)
-                .ok_or(Error::CannotRoute);
-            best_section.map(|(len, connected)| {
-                (
-                    len,
-                    connected
-                        .into_iter()
-                        .sorted_by(|lhs, rhs| target_name.cmp_distance(lhs, rhs)),
-                )
-            })
+
+            let mut dg_size = 0;
+            let mut node_to_send = Vec::new();
+            for (idx, (prefix, len, connected)) in filtered_sections.enumerate() {
+                node_to_send.extend(connected.into_iter());
+                dg_size = delivery_group_size(len);
+
+                if &prefix == self.our_prefix() {
+                    // Send to all connected targets so they can forward the message
+                    node_to_send.retain(|&x| x != *self.our_id().name());
+                    dg_size = node_to_send.len();
+                    break;
+                }
+                if idx == 0 && node_to_send.len() >= dg_size {
```

Comment:
Does it make sense to check the index here? I mean, if the number of candidate nodes exceeds `dg_size`, then we're fine regardless of whether that's the first section on the list or not, aren't we?

Reply:
I guess this would be more efficient (do we need it though since that should rarely happen / does not seem to impact test performance as well).
But we would have to be careful, i.e we cannot just remove `idx == 0`, because if we use a delivery group, the node to send need to only contain nodes from a single section.

In this case we would need to first check:
```rust
let group_dg_size = delivery_group_size(len);
let is_our_prefix = &prefix == self.our_prefix();
if !is_our_prefix && connected.len() >= group_dg_size {
     // Only send to that group since it is forward progress.
     node_to_send = connected;
     dg_size = delivery_group_size(group_dg_size);
     break;
}
if is_our_prefix {
     // Send to all connected targets so they can forward the message
     node_to_send.retain(|&x| x != *self.our_id().name());
     dg_size = node_to_send.len();
     break; 
}
```

I'm not 100% sure it is simpler/better.

In the current implementation, we either have the normal case where we send to delivery group from 3 peers, and if one of them is not connected enough it will send to everyone closer that them + own section (yes, that can be quite spammy)

In the alternative, we either have the normal case where we send to delivery group from 3 peers, and if one of them is not connected enough it will send to another section that is closer but connected enough, and if none such exist, it will send to everyone closer than them + own section (yes, that would be the same spammy, but in fewer cases).

Is that what you had in mind?

Reply:
I think I still favor the current implementation.

Reply:
> it will send to everyone closer that them + own section (yes, that can be quite spammy)
Ah, I missed this part. Yeah, that is spammy, but it should be an unusual case, so maybe not that big of a problem.
Anyway, it makes more sense now, so I'm okay with leaving it as is :+1: